zink/codegen/src/event.rs

```diff
@@ -1,19 +1,339 @@
+            }
+
+            let field_count = fields.unnamed.len();
+            let field_bindings = (0..field_count)
+                .map(|i| quote::format_ident!("v{}", i))
+                .collect::<Vec<_>>();
+            let ref_patterns = field_bindings
+                .iter()
+                .map(|id| quote!(ref #id))
+                .collect::<Vec<_>>();
+
+            let field_types: Vec<_> = fields.unnamed.iter().map(|f| &f.ty).collect();
+            validate_types(&field_types)?;
+
+            // Generate the appropriate log call based on field count
+            let log_impl = match field_bindings.as_slice() {
+                [] => quote! {
+                    unsafe {
+                    zink::ffi::evm::log0(stringify!(#variant_name).as_bytes());
+                    }
+                },
+                [v0] => quote! {
+                    let topic1 = #v0.bytes32();
+                    unsafe {
+                    zink::ffi::evm::log1(topic1, stringify!(#variant_name).as_bytes());
+                    }
+                },
+                [v0, v1] => quote! {
+                    let topic1 = #v0.bytes32();
+                    let topic2 = #v1.bytes32();
+                    unsafe {
+                    zink::ffi::evm::log2(topic1, topic2, stringify!(#variant_name).as_bytes());
+                    }
+                },
+                [v0, v1, v2] => quote! {
+                    let topic1 = #v0.bytes32();
+                    let topic2 = #v1.bytes32();
+                    let topic3 = #v2.bytes32();
+                    unsafe {
+                    zink::ffi::evm::log3(topic1, topic2, topic3, stringify!(#variant_name).as_bytes());
+                    }
+                },
+                [v0, v1, v2, v3] => quote! {
+                    let topic1 = #v0.bytes32();
+                    let topic2 = #v1.bytes32();
+                    let topic3 = #v2.bytes32();
+                    let topic4 = #v3.bytes32();
+                    unsafe {
+                    zink::ffi::evm::log4(topic1, topic2, topic3, topic4, stringify!(#variant_name).as_bytes());
+                    }
+                },
+                _ => unreachable!(),
+            };
+
+            Ok(quote! {
+                #enum_name::#variant_name(#(#ref_patterns),*) => {
+                    #log_impl
+                }
+            })
+        }
+        Fields::Unit => Ok(quote! {
+            #enum_name::#variant_name => {
+                unsafe {
+                zink::ffi::evm::log0(stringify!(#variant_name).as_bytes())
+                }
+            }
+        }),
+    }
+}
+
+/// Validate field types
+fn validate_types(types: &[&Type]) -> Result<()> {
+    for ty in types {
+        if !is_supported_type(ty) {
+            return Err(Error::new_spanned(
+                ty,
+                format!("Unsupported type for event field: {}", quote!(#ty)),
+            ));
+        }
```

Comment:
matches!(sol_abi::Param::from(&ty.to_string()), )
```suggestion
       if matches!(sol_abi::Param::from(&ty.to_string()), sol_abi::Param::Unknown(_)) {
            return Err(Error::new_spanned(
                ty,
                format!("Unsupported type for event field: {}", quote!(#ty)),
            ));
       }
```